In [1]:
import pandas as pd
# from scipy.stats import chi2_contingency
import sys
sys.path.append('../')
from src.stats_utils import analyze_independence

df = pd.read_csv("../data/1_independencia_error_vs_clase.csv")

observed, expected, chi2, p_value = analyze_independence(df, "predicted_class", "error_type")

print(f"Chi-Square Statistic: {chi2:.4f}")
print(f"P-value: {p_value:.4e}")



Chi-Square Statistic: 37.9521
P-value: 5.7386e-09


Chi-Square Statistic (37.9521): This is the calculated mathematical distance between reality (your model's actual errors) and theory (how the errors would look if the model were perfectly unbiased and errors were distributed evenly). A score of 0 means perfect independence. A score of 38 is a massive deviation.

P-value (5.7386e-09): This is the probability that a deviation of 37.95 occurred by pure random chance. 5.73e-09 is 0.00000000573. It is effectively zero, falling far below the standard 0.05 threshold.

The analytical verdict: You reject the null hypothesis. The type of error your model makes is inherently dependent on the class it is predicting. As we analyzed earlier by looking at the raw matrix, your model has a severe algorithmic bias—specifically, it is aggressively over-predicting Class C and generating an abnormal amount of False Positives for that category.

In [9]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 600 entries, 0 to 599
Data columns (total 2 columns):
 #   Column           Non-Null Count  Dtype
---  ------           --------------  -----
 0   predicted_class  600 non-null    str  
 1   error_type       600 non-null    str  
dtypes: str(2)
memory usage: 9.5 KB
